### DISCUSSION TITLE
Automatic mapping feature? (Just walk around, or import from robot vac/image?)

### DISCUSSION THREAD

**Discussion by scrampker** [ESPresense/ESPresense-companion#822] [Ideas]
Maybe I'm naive, but most other sensors/equipment I own similar-ish to this has some sort of self-mapping system. Something along the lines of walking the perimeter of a room.

I could imagine a simple UI that lets me pick the tracking device, then create rooms as I go. It should be possibly to stitch the rooms together with fairly straightforward logic, unless I'm missing something here.

The biggest barrier to entry with this is building out complicated blueprints for my house, and not being able to tweak it easily in the plugin.

If that's too hard to implement, perhaps being able to import a simplified screenshot of my house from my robot vacuum would be possible? The only way I can think of right now would be to print out that robot map and tape it on the monitor.

**Comment by dxmnkd316** [comment]
I'll be honest, I don't think even with 50+ base stations you're going to get the accuracy you are looking for to be able to just map the house. The accuracy in my house with 10 stations and extensive calibration is probably in the range of 1m. Which is still pretty good using RSSI which is highly susceptible to obstructions like a fridge, a computer monitor, or even the orientation of your phone. 

I think the straightforward logic isn't very straightforward. That's a pretty complex algorithm that companies like iRobot and the like have spent ages working out. And they're still not that great.

I spent an afternoon with a tape measure and a calculator to get the dimensions of my house and plugged them into the coordinates in the config. I suppose it's possible your house is large and complex, but it shouldn't take too much more than a couple hours at most. You're still going to need to break out a tape measure and get the exact location of the base stations anyways. 

Just my $0.02

**Reply by scrampker** [reply]
It just seems like it would be pretty easy to initiate a learn mode, walk the perimeter of a room, and the software could record the signal from every base station. Absolute position irrelevant, you'd at least have all possible combinations of signal that equal being in that room. No matter what fridges or interference might exist. You'd basically be doing the calibration from the get-go.

**Reply by RN-Say** [reply]
Magicplan (among several similar apps) can do this with your phone camera pretty quickly and with surprising accuracy. Trying to build a model from radio waves would be hundreds of times more difficult and give a less accurate result imo. 

A learn mode is a pretty interesting idea but I can imagine that the profile that you built could change dramatically from hour to hour.

**Reply by dxmnkd316** [reply]
> Magicplan (among several similar apps) can do this with your phone camera pretty quickly and with surprising accuracy. Trying to build a model from radio waves would be hundreds of times more difficult and give a less accurate result imo.
> 
> A learn mode is a pretty interesting idea but I can imagine that the profile that you built could change dramatically from hour to hour.

I have a spreadsheet I use to optimize the node settings using thousands of datapoints collected over hours. Some are fairly typical looking bell curves. But they're not consistent. They're either long-tailed or among those that are Gaussian, they have a fairly wide range of standard deviations. Some are very tight distributions others are flat. 

They are also quite dependent on interferences. Appliances and monitors are quite good at blocking signals, for example. When my monitor blocks a direct line-of-sight to the node five feet away, it can often be read as 10-15 feet away.

all of this is to say there's an inherent fuzziness in using RF. Which is why accuracy in practice scales with number of nodes.  Your point about magic plan is a good one. It's much better, but still not as good as LiDAR at generating room dimensions. Maybe MP can take advantage of some of the feature of higher end devices that have LiDAR built in.
